Support to build tests in Unix

jyoungyun opened the pull request:
Support to build tests in Unix

Fix #4437

Support to build tests in Unix.
But this option still turns off because of some build failure issues.
The build failure issues will be fixed on next step.

Added tests options.
It also provides to pass some tests option like priority,
gcstresslevel, sequential and ilasmroundtrip.

Do not import mono.targets and roslyn.xplat.targets.
Unix can not find these targets in this project. It seems not to be used anymore.

DebugType and DebugSymbols are disabled.
Roslyn does not support writing PDBs on Unix. To disable these properties in
whole projects, buildtest.sh script will pass below options directly.
/p:DebugSymbols=false /p:DebugType=none

By not supporting CodeTaskFactory in Linux .NET Core, do not use it.
It does not affect the real operation of tests running.

Signed-off-by: Jiyoung Yun jy910.yun@samsung.com

jyoungyun commented:
The related issue: #4860

\cc:@leemgs @myungjoo @mylibero @gkhanna79 @jkotas

myungjoo commented:
Great! Just be careful that Linux is not the only platform other than Windows in this project. If it's too Linux-specific, you may need to restrict the execution condition.

janvorli commented on src/pal/tools/gen-buildsys-clang.sh:
```diff
@@ -131,16 +133,16 @@ if [[ -n "$CROSSCOMPILE" ]]; then
     fi
-    cmake_extra_defines="$cmake_extra_defines -C $1/cross/$build_arch/tryrun.cmake"
-    cmake_extra_defines="$cmake_extra_defines -DCMAKE_TOOLCHAIN_FILE=$1/cross/$build_arch/toolchain.cmake"
+    cmake_extra_defines="$cmake_extra_defines -C $basePath/../../../cross/$build_arch/tryrun.cmake"
```
I don't like the script depending on its location. What was wrong with the path being passed in as a parameter?

janvorli commented:
This change has a lot of duplication between build.sh and buildtest.sh, so maintaining those two in the future would be error prone and so it would be difficult to maintain. The same for the CMakeLists.txt in the tests and the root one.
I would rather prefer model where we build the tests using the root build.sh, using the existing "skiptests" to exclude the build and when the tests/buildtests.sh builds just the managed part of the tests and is invoked from the build.sh and not meant to be run standalone => it would get the necessary parameters from the build.sh without the need to parse the command line (the parameters would have fixed positions, like we have it for the gen-buildsys-clang.sh.

jyoungyun commented:
I understand your purpose. I've thought the test/buildtest.sh script should be run standalone without build.sh. If it uses build.sh to build native tests, the code is more simple maybe. Besides, the gen-buildsys-clang.sh does not need to fix as you said. I will update patch soon.

jyoungyun commented:
@janvorli @gkhanna79 By not supporting CodeTaskFactory in .NET Core, I need other way to replace CodeTaskFactory for generating param lists. I have a plan to create external dll file likes Tools.dll before running build.proj and import it using UsingTask when the OSEnvironment is Linux. Is there any better idea for it ? If not, I will update it soon.

gkhanna79 commented on tests/CMakeLists.txt:
```diff
@@ -1,11 +1,44 @@
 cmake_minimum_required(VERSION 2.8.12)
 
+function(clr_unknown_arch)
+	if (WIN32)
```
For win32, we support building tests for x86 on Windows as well.

jyoungyun replied:
This phrase will be removed as per @janvorli comments.

jyoungyun commented:
@janvorli @gkhanna79 I added new task assembly for generating param lists in tests/src/Common/CLRTestTools/ directory. Please check the detail code and if these codes are unfit in the light of the CoreCLR concept, let me know it. (This commit is not the final version yet, but it covers a lot of thing to build the unix-tests in Linux.)

janvorli commented:
@gkhanna79 can you please comment on the task? This is out of my knowledge.

gkhanna79 commented:
@jyoungyun Please have a look at https://github.com/gkhanna79/coreclr/tree/XPlatTests that I did couple months ago before I paused it. I dont believe you need to make CodeTaskFactory changes.

Please go over these changes and we can then discuss the next steps if you are keen to get things working.

dnfclas commented:
__@jyoungyun__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

jyoungyun commented:
@gkhanna79 I saw your codes. The progress is almost the same to me. Your code does not use GenerateParamList task in bash script, right ? If do that, I don't need to make a special task for CodeTaskFactory. I modified build.sh like you without adding new script (tests/buildtests.sh). Instead I added an additional option(testsonly). It will be helpful to build tests only.
Actually, this patch still throws some error messages.
```
/home/jyoung/git/dotnet/coreclr/tests/src/CLRTest.Execute.Bash.targets(29,7): error MSB4184: The expression "[MSBuild]::MakeRelative(/...(SKIP).../ThreadStartBool_1/, ...(SKIP).../ThreadStartBool/ThreadStartBool.sh)" cannot be evaluated. Invalid URI: The format of the URI could not be determined. 
```
But as I know this is the known issue. (https://github.com/dotnet/corefx/issues/1745)
```
/home/jyoung/git/dotnet/coreclr/tests/src/dir.targets(119,6): error : The native project files are missing in  /home/jyoung/git/dotnet/coreclr/bin/tests/obj/Linux.arm.Debug/Native/src/Exceptions/ForeignThread/Debug/ please run build from the root of the repo at least once
```
```
eh01.cs(34,20): error CS0012: The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'mscorlib,
```
I want to fix these issues on next step. Please check the detail codes and comments here. Thanks.
(I think the skiptests option is really useful until now. Because to build the tests subdirectory takes a lot of time. In my PC(my PC is very slow...), full build including tests took around 8 hours!!!!.)

leemgs commented:
> I think the skiptests option is really useful until 
> now. Because to build the tests subdirectory
> takes a lot of time...
Actually, we have been improving the development productivity thanks to the 'skiptests' option.

jyoungyun commented:
@gkhanna79 Could you help me? When I tried to build it, the operation ended successfully without  below error. I can't find points to make build break in my patch. 

> MSBuild Error
>
> error MSB4186: Invalid static method invocation syntax: "[System.IO.Path]::GetFullPath()". Method 'System.IO.Path.GetFullPath' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). 
Indication 2 
'''

jyoungyun commented:
@gkhanna79 Ready for review. Please check it and let me know if there is any issues.

myungjoo commented:
Acked-By: MyungJoo Ham

jasonwilliams200OK commented:
Thanks for doing this @jyoungyun! IMO, at least one long-running CI job (the machine where no Mono PCL reference assemblies are available, with bare minimum operating system libs) should be configured to build these tests along with everything (full stack), so we can continuously validate that nothing is broken. With current master, when building the tests on Windows, I run into errors like:

> CSC : error CS0006: Metadata file 'C:\Users\Peter\Source\Repos\coreclr\tests\..\bin\Product\Windows_NT.x64.Debug\ref\m
scorlib.dll' could not be found [C:\Users\Peter\Source\Repos\coreclr\tests\src\JIT\Methodical\xxobj\operand_speed_dbgr
efanyval.csproj]

jyoungyun commented:
@jasonwilliams200OK Yes, I'm for your comments. But currently to build tests binaries outside Windows still has some problem. So after all building is clear, the option will turn on by default.
When you built this patch on your Windows machine, did you get the error message ? Could you give me more explanation please ?

jasonwilliams200OK commented:
@jyoungyun, I built the dotnet/coreclr master branch (for Gentoo Linux testing) using `tests\buildtest.cmd clean` command as documented here: https://github.com/dotnet/coreclr/blob/e5ee95e/Documentation/building/unix-test-instructions.md#coreclr-tests. At which point, I received those (`CS0006`) errors.

jyoungyun commented:
@jasonwilliams200OK You need to build coreclr first before building test cases. Try `build.cmd clean` at the coreclr root directory. This command makes to build whole thing including test case. After that, you can build test case only. In my case, it builds successfully using the latest coreclr master branch(commit:38a8915).

jasonwilliams200OK commented:
@jyoungyun, so if I run `buildtest.cmd clean` right after cloning CoreCLR on Windows, it will not work?

> Try `build.cmd clean` at the coreclr root directory. This command makes to build whole thing including test case. After that, you can build test case only

If `build.cmd clean` command builds everything including tests and this command is a prerequisite, why would we run `tests\buildtest.cmd clean` command after already building everything?

The errors I am seeing is *I think* because I have VS2015 with Xamarin stuff installed. I will try on a separate VM to see if I can reproduce this issue. But I am very sure that running `tests\buildtest.cmd clean` on a fresh clone copy should work as the docs are suggesting (without building the whole CoreCLR), and I have successfully built those tests number of times in past.

jyoungyun commented:
@jasonwilliams200OK 

> If build.cmd clean command builds everything including tests and this command is a prerequisite, why would we run tests\buildtest.cmd clean command after already building everything?

I mean first time you can build whole thing(at this time you don't need to build test cases again because as you said the test case was already built by build.cmd clean) and next time if you want to get new tc sets, you can build tests only without building the whole CoreCLR. As per your error message, it couldn't find the ref/mscorlib.dll and it created by building coreclr. But as you said you have successfully built tests only without building the whole CoreCLR, then my comments might be wrong. I will try to install VS2015 with Xamarin stuff and then update the build status again.

jyoungyun commented:
@jasonwilliams200OK When I built the test case without building the whole CoreCLR, I faced the same erros like you. So I think you need to build the whole CoreCLR first.

> CSC : error CS0006: 'c:\git\dotnet\coreclr\tests\\..\bin\Product\Windows_NT.x64
.Debug\ref\mscorlib.dll' could not be found.  [c:\git\dotnet\core
clr\tests\src\

mkborg commented:
By default `build.cmd` builds almost everything (including tests marked with priority `0`).

`tests\buildtest.cmd` can be necessary for non-default tasks:

- `tests\buildtest.cmd priority 1` can be used to build additional (almost 3 thousand) tests
- `tests\buildtest.cmd clean` cleans changes made during tests execution (`tests\runtest.cmd`) and is necessary to re-run tests

For example, usually I do the following:
```
build.cmd
tests\buildtest.cmd priority 1
tests\runtest.cmd
```
but sometimes:
```
build.cmd
tests\runtest.cmd
tests\buildtest.cmd priority 1 clean
tests\runtest.cmd
```

jasonwilliams200OK commented:
>  I think you need to build the whole CoreCLR first.

then I must have mixed up with something: I wrote build steps to build CoreCLR on FreeBSD, NetBSD, Gentoo and Alpine (which is currently work in progress) starting from configuring VM on Windows 10 Hyper-V to running `./build.sh`, but none of those included CoreFX testing through CoreCLR, I have separate CoreFX build recipes in which test are included. I think I will go through them again and maybe just try building CoreFX straightaway and run CoreFX test on Gentoo.

mkborg commented:
@jyoungyun I suppose this change is applicable to Linux then how to activate build of the tests?
> coreclr/build.sh x64 debug verbose
> ...
> **Skipping** tests build.

jyoungyun commented on build.sh:
```diff
@@ -451,6 +520,8 @@ __SkipRestore=""
 __SkipCoreCLR=0
 __SkipMSCorLib=0
+# The default value of __SkipTests is off temporarily.
+__SkipTests=1
```
@mkborg Currently there is no option to enable the test building in build.sh scripts. Instead you can enable it manully by updating __SkipTests=0. Until now, there are some build error issue in test building so I couldn't enable the option yet. After fixing whole issues, I will enable. Apart from this, I thought about adding new option like testsOnly to enable only tests build but as I know tests binaries are very relevant to coreclr libraries. So it didn't add it.

mkborg commented:
@jyoungyun May be we shall introduce `tests/buildtest.sh` (similar to `tests/buildtest.cmd`)?

jyoungyun commented:
@mkborg I already tried to create `tests/buildtest.sh` script. But other maintainers(@janvorli and @gkhanna79 ) want to use build.sh without new scripts. As I know, the `test/buildtest.cmd` file does not mean to be run standalone. It's only for other option of test building like priority level, sequential, gcsimulator and so on. So they want to follow the same rule in this case too.

@janvorli 's comment:
> This change has a lot of duplication between build.sh and buildtest.sh, so maintaining those two in the future would be error prone and so it would be difficult to maintain. The same for the CMakeLists.txt in the tests and the root one.
I would rather prefer model where we build the tests using the root build.sh, using the existing "skiptests" to exclude the build and when the tests/buildtests.sh builds just the managed part of the tests and is invoked from the build.sh and not meant to be run standalone

mkborg commented:
To prepare CoreCLR for ARM we need:
- native ARM code (cross build on powerful x86_64 server is much faster then native build on ARM board)
- managed code (I doubt if it is possible or feasible on ARM board)

So we may need options for build.sh to control each part of the build.

mkborg commented:
@jyoungyun Right now I am trying to build tests on Linux but it is very slow because it executes only single 'corerun.exe' and uses only single CPU. Is it possible to make it parallel to use all CPUs?

jasonwilliams200OK commented on tests/src/CLRTest.Execute.Bash.targets:
```diff
@@ -163,7 +163,7 @@ fi
         if [ ! -f "$_DebuggerFullPath" ]
         then
-            echo The Debugger FullPath \"$_DebuggerFullPath\" doesn\'t exist
+            echo "The Debugger FullPath "$_DebuggerFullPath" doesn't exist"
```
nit: should we surround `$_DebuggerFullPath` with single quote?

jyoungyun replied:
@jasonwilliams200OK You're right, it does not need the single quote. I will fix it soon. But I want to review from maintainers first. This patch is been too long....

jyoungyun commented:
@janvorli 
`/home/janvorli/git/coreclr/tests/src/CLRTest.Execute.Batch.targets(25,7): error MSB4184: The expression "[MSBuild]::MakeRelative(/home/janvorli/git/coreclr/tests/../bin/tests/Linux.x64.Release/baseservices/threading/paramthreadstart/ThreadStartBool_1/, /home/janvorli/git/coreclr/tests/../bin/tests/Linux.x64.Release/baseservices/threading/paramthreadstart/ThreadStartBool/ThreadStartBool.cmd)" cannot be evaluated. Invalid URI: The format of the URI could not be determined. [/home/janvorli/git/coreclr/tests/src/baseservices/threading/paramthreadstart/ThreadStartBool_1.csproj]`
I saw the same issue and as I know it is a known issue in corefx(https://github.com/dotnet/corefx/issues/1745).
And if you are said about below issue, I saw the issue too.
`eh01.cs(34,20): error CS0012: The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'mscorlib,`
`/home/jyoung/git/dotnet/coreclr/tests/src/dir.targets(119,6): error : The native project files are missing in  /home/jyoung/git/dotnet/coreclr/bin/tests/obj/Linux.arm.Debug/Native/src/Exceptions/ForeignThread/Debug/ please run build from the root of the repo at least once`
Actually I wanna fix whole issues which are related to build tests in Linux at a time, but I think it takes a long time and I look to others for help to fix the remain issues. So I want to merge it in the first step and continue on it. Thanks for your reply!

gkhanna79 commented:
The reference to ".cmd" does not sounds right, it should refer to ".sh". Also, the native test pieces, for non-Windows, are built as part of the product build on non-Windows. We should consider moving them out of the product build into the test build and then figure out what is missing.

danmosemsft commented:
@jyoungyun do you expect to have a chance to rebase this and address feedback ?

jyoungyun commented:
@danmosemsft Isn't there the same or similar commit in progress? If there is no plan about it, I want to rebase and keep doing it.

gkhanna79 commented:
Please continue to rebase and work on this.

danmosemsft commented:
@jyoungyun do you expect to be able to continue this PR?

danmosemsft commented:
@jyoungyun?

jkotas commented:
Closing old stale PRs. @jyoungyun feel free to open a new PR once you get a chance to work on this again.
